wagtail/permission_policies/pages.py

```diff
@@ -0,0 +1,216 @@
+from django.contrib.auth import get_user_model
+from django.db.models import (
+    CharField,
+    Exists,
+    F,
+    OuterRef,
+    PositiveIntegerField,
+    Q,
+    Value,
+)
+
+from wagtail.models import GroupPagePermission, Page
+from wagtail.permission_policies.base import BasePermissionPolicy
+
+
+class PagePermissionPolicy(BasePermissionPolicy):
+    perm_cache_name = "_page_perm_cache"
+    _explorable_root_instance_cache_name = "_explorable_root_instance_cache"
+    _explorable_instances_cache_name = "_explorable_instances_cache"
```

Comment:
Not a big fan of these two, honestly. Caching the permission objects via `perm_cache_name` makes sense as we try to replicate how Django caches the permission objects for users in the default `ModelBackend` authentication backend. However, I don't think we want to stretch this to the caching of other objects as well.

#9594 has a different approach where the `UserPagePermissionsProxy` object itself is cached, and query results are cached using `@cached_property`.

I thought about doing the same in here, e.g. allowing `user` to be passed to the constructor and making sure only one instance of `PagePermissionPolicy` is created for one user. However, the design of permission policies is more general, as it also has use cases that are not specific to a user.

It seems we still need to have _somewhere_ to cache these user-specific queries, which seems to be what `UserPagePermissionsProxy` was for. That said, I'd much prefer this cache object to live somewhere closer to `PagePermissionPolicy` than in `models/__init__.py`.

Reply:
I tried with the approach similar to #9594 (see https://github.com/laymonage/wagtail/commit/e8e3928cbd0feb06c58c1eccb54289de0c963674), but as discussed, it doesn't seem to be a pattern that we want to do.

I've updated this to only cache the explorable root instance.